metric_learn/constraints.py

```diff
@@ -33,6 +35,110 @@ def positive_negative_pairs(self, num_constraints, same_length=False,
+        Input data, where each row corresponds to a single instance.
+      k_genuine : int
+        Number of neighbors of the same class to be taken into account.
+      k_impostor : int
+        Number of neighbors of different classes to be taken into account.
+
+    Returns
+    -------
+    triplets : array-like, shape=(n_constraints, 3)
+      2D array of triplets of indicators.
+    """
+
+    labels, labels_count = np.unique(self.partial_labels, return_counts=True)
+    n_labels = len(labels)
+    len_input = np.size(self.partial_labels, 0)
+
+    # Handle the case where there are too few elements to yield k_genuine or
+    # k_impostor neighbors for every class.
+
+    k_genuine_vec = np.ones(n_labels, dtype=np.intp)*k_genuine
+    k_impostor_vec = np.ones(n_labels, dtype=np.intp)*k_impostor
+
+    for i in range(n_labels):
+      if (k_genuine + 1 > labels_count[i]):
+        k_genuine_vec[i] = labels_count[i]-1
+        warnings.warn("The class {} has {} elements, which is not sufficient "
+                      "to generate {} genuine neighbors as specified by "
+                      "k_genuine. Will generate {} genuine neighbors instead."
+                      "\n"
+                      .format(labels[i], labels_count[i], k_genuine+1,
+                              k_genuine_vec[i]))
+      if (k_impostor > len_input - labels_count[i]):
+        k_impostor_vec[i] = len_input - labels_count[i]
+        warnings.warn("The class {} has {} elements of other classes, which is"
+                      " not sufficient to generate {} impostor neighbors as "
+                      "specified by k_impostor. Will generate {} impostor "
+                      "neighbors instead.\n"
+                      .format(labels[i], k_impostor_vec[i], k_impostor,
+                              k_impostor_vec[i]))
+
+    triplets = np.empty((np.dot(k_genuine_vec*k_impostor_vec, labels_count),
+                         3), dtype=np.intp)
+
+    start = 0
+    finish = 0
+    neigh = NearestNeighbors()
+
+    for i in range(n_labels):
+
+        # generate mask for current label
+        gen_mask = self.partial_labels == labels[i]
+        gen_indx = np.where(gen_mask)
+
+        # get k_genuine genuine neighbours
+        neigh.fit(X=X[gen_indx])
+        gen_neigh = np.take(gen_indx, neigh.kneighbors(
+                            n_neighbors=k_genuine_vec[i],
+                            return_distance=False))
+
+        # generate mask for impostors of current label
+        imp_indx = np.where(np.invert(gen_mask))
+
+        # get k_impostor impostor neighbours
+        neigh.fit(X=X[imp_indx])
+        imp_neigh = np.take(imp_indx, neigh.kneighbors(
+                            n_neighbors=k_impostor_vec[i],
+                            X=X[gen_mask],
+                            return_distance=False))
+
+        # length = len_label*k_genuine*k_impostor
+        finish += labels_count[i]*k_genuine_vec[i]*k_impostor_vec[i]
```

Comment:
Let's compute this outside the loop: `finish_indices = labels_count * k_genuine_vec * k_impostor_vec`

Reply:
I calculated directly the cumulative sum outside the for loop, take a look at it and let me know what you think.